crates/bevy_ptr/src/batch.rs

```diff
@@ -0,0 +1,282 @@
+use crate::ThinSimdAlignedSlicePtr;
+
+use core::{cell::UnsafeCell, marker::PhantomData};
+
+//Required for as_inner* conversions
+//Re-exported from the [`bytemuck`] crate to ensure users can implement this trait for their own types
+pub use bytemuck::TransparentWrapper;
+
+/*
+NOTE: We define this constant here as both [`bevy_ptr`] and [`bevy_ecs`] need to know about it.
+
+If this is a problem, this can be replaced with code that looks something like the following:
+
+    #[cfg(all(any(target_feature = "avx"), not(target_feature = "avx512f")))]
+    pub const MAX_SIMD_ALIGNMENT: usize = 32;
+
+    #[cfg(any(target_feature = "avx512f"))]
+    pub const MAX_SIMD_ALIGNMENT: usize = 64;
+
+    //All platforms get 16-byte alignment on tables guaranteed.
+    #[cfg(not(any(target_feature = "avx512f")))]
+    pub const MAX_SIMD_ALIGNMENT: usize = 16;
+*/
+/// The maximum SIMD alignment for a given target.
+/// MAX_SIMD_ALIGNMENT is 64 for the following reasons:
+///  1. This ensures that table columns are aligned to cache lines on x86
+///  2. 64 is the maximum alignment required to use all instructions on all known CPU architectures.
+///     This simplifies greatly handling cross platform alignment on a case by case basis; by aligning to the worst case, we align for all cases
+///  3. The overhead of aligning columns to 64 bytes is very small as columns will in general be much larger than this
+pub const MAX_SIMD_ALIGNMENT: usize = 64;
+
+//TODO: AoSoA representations
+
+//TODO: when possible, compute alignments automatically using the GCD (requires generic const expressions) and ensure the
+//batch lambda accepts arguments of AlignedBatchTrait<T, N> directly. This will let the alignment be automatically computed
+//and allow different query elements to have different alignments.
+
+/// This trait allows generic code using [`AlignedBatchT`] with varying sizes and alignments
```

Comment:
This needs doc comments explaining why it's a sealed trait, with links to reading material about the pattern.

Reply:
Just a quick question before I add this in -- `bevy_ptr` has a sealed `UnsafeCellDeref` trait in the root, should the documentation on sealed traits be linked there too?

Reply:
Yes, but make a new issue for that please.